fsm/fsm.go

```diff
@@ -72,19 +78,35 @@ func NewFSMHandler(parameters Parameters) (statemachine.StateHandler, error) {
 // At the end it executes the specified handler for the final state,
 // if specified
 func (d fsmHandler) Plan(events []statemachine.Event, user interface{}) (interface{}, uint64, error) {
+	userValue := reflect.ValueOf(user)
+	if d.reachedFinalityState(userValue.Elem().Interface()) {
+		d.eventProcessor.ClearEvents(events, statemachine.ErrTerminated)
+		return nil, uint64(len(events)), statemachine.ErrTerminated
+	}
 	eventName, err := d.eventProcessor.Apply(events[0], user)
 	if err != nil {
 		log.Errorf("Executing event planner failed: %+v", err)
 		return nil, 1, nil
 	}
-	userValue := reflect.ValueOf(user)
 	currentState := userValue.Elem().FieldByName(string(d.stateKeyField)).Interface()
 	if d.notifier != nil {
 		d.notifier(eventName, userValue.Elem().Interface())
 	}
+	_, final := d.finalityStates[currentState]
+	if final {
+		d.eventProcessor.ClearEvents(events[1:], statemachine.ErrTerminated)
+		return nil, uint64(len(events)), statemachine.ErrTerminated
```

Comment:
Are state entry functions not supported on terminal states?  If so it should be documented.